### PR TITLE
implement a streams map for handling unidirectional and bidirectional streams

### PR DESCRIPTION
**Pull request by marten-seemann**
Work towards #996.

### CODE REVIEW

**Comment by marten-seemann** [comment]
We could also use the new stream maps for incoming and outgoing streams for gQUIC. @lucas-clemente, WDYT?

File: streams_map.go [hunk 1]
```diff
@@ -1,234 +1,175 @@
...
+func (m *streamsMap) getStreamType(id protocol.StreamID) streamType {
+	if m.perspective == protocol.PerspectiveServer {
+		switch id % 4 {
+		case 0:
+			return streamTypeIncomingBidi
+		case 1:
+			return streamTypeOutgoingBidi
+		case 2:
+			return streamTypeIncomingUni
+		default: // case 3, have to use default so Go recognizes the end of function
```
Comment (lucas-clemente) [inline streams_map.go]: case 3: and panic in default is the right pattern here.

File: streams_map.go [hunk 2]
```diff
@@ -1,234 +1,175 @@
...
+		}
+	} else {
+		switch id % 4 {
+		case 0:
+			return streamTypeOutgoingBidi
+		case 1:
+			return streamTypeIncomingBidi
+		case 2:
+			return streamTypeOutgoingUni
+		default: // case 3, have to use default so Go recognizes the end of function
```
Comment (lucas-clemente) [inline streams_map.go]: same.

File: streams_map.go [hunk 3]
```diff
@@ -1,234 +1,175 @@
...
-		m.openStreamOrErrCond.Wait()
+func (m *streamsMap) DeleteStream(id protocol.StreamID) error {
+	switch m.getStreamType(id) {
+	case streamTypeIncomingBidi:
+		return m.incomingBidiStreams.DeleteStream(id)
+	case streamTypeOutgoingBidi:
+		return m.outgoingBidiStreams.DeleteStream(id)
+	case streamTypeIncomingUni:
+		return m.incomingUniStreams.DeleteStream(id)
+	default: // case streamTypeOutgoingUni:
```
Comment (lucas-clemente) [inline streams_map.go]: same

File: streams_map.go [hunk 4]
```diff
@@ -1,234 +1,175 @@
...
-		m.nextStreamOrErrCond.Wait()
+func (m *streamsMap) GetOrOpenReceiveStream(id protocol.StreamID) (receiveStreamI, error) {
+	switch m.getStreamType(id) {
+	case streamTypeOutgoingBidi:
+		return m.outgoingBidiStreams.GetStream(id)
+	case streamTypeIncomingBidi:
+		return m.incomingBidiStreams.GetOrOpenStream(id)
+	case streamTypeIncomingUni:
+		return m.incomingUniStreams.GetOrOpenStream(id)
+	default: // streamTypeOutgoingUni
```
Comment (lucas-clemente) [inline streams_map.go]: same

File: streams_map.go [hunk 5]
```diff
@@ -1,234 +1,175 @@
...
-		return errMapAccess
+func (m *streamsMap) GetOrOpenSendStream(id protocol.StreamID) (sendStreamI, error) {
+	switch m.getStreamType(id) {
+	case streamTypeOutgoingBidi:
+		return m.outgoingBidiStreams.GetStream(id)
+	case streamTypeIncomingBidi:
+		return m.incomingBidiStreams.GetOrOpenStream(id)
+	case streamTypeOutgoingUni:
+		return m.outgoingUniStreams.GetStream(id)
+	default: // streamTypeIncomingUni
```
Comment (lucas-clemente) [inline streams_map.go]: same

File: streams_map.go [hunk 6]
```diff
@@ -1,234 +1,175 @@
...
-	closeErr           error
-	nextStreamToAccept protocol.StreamID
+	sender            streamSender
+	newFlowController func(protocol.StreamID) flowcontrol.StreamFlowController
 
-	newStream newStreamLambda
+	outgoingBidiStreams *outgoingBidiStreamsMap
+	outgoingUniStreams  *outgoingUniStreamsMap
+	incomingBidiStreams *incomingBidiStreamsMap
+	incomingUniStreams  *incomingUniStreamsMap
```
Comment (lucas-clemente) [inline streams_map.go]: Some of the functions here could be simpler if you made this a map[streamType]commonInterface – just a thought if that didn't occur to you.

Comment (marten-seemann) [reply streams_map.go]: There no overlapping interface between incoming and outgoing stream maps.

Comment (lucas-clemente) [reply streams_map.go]: DeleteStream is common. Not important though.

**Pull request merged**